qutebrowser/utils/objreg.py

```diff
@@ -38,6 +39,46 @@ class CommandOnlyError(Exception):
     """Raised when an object is requested which is used for commands only."""
 
 
+_WindowRefType = weakref.ReferenceType['mainwindow.MainWindow']
+
+
+class WindowAccessHistory():
+    """Contains an ordered list of windows, sorted by last added.
+
+    Uses weak references to avoid memory leaks.
+    """
+
+    def __init__(self) -> None:
+        self._windows: list[_WindowRefType] = []
+
+    def append(self, window: 'mainwindow.MainWindow') -> None:
+        """Append a window to the access history."""
+        self.remove(window)
+        self._windows.append(weakref.ref(window))
+
+    def remove(self, window: 'mainwindow.MainWindow') -> None:
+        """Remove a window from the access history."""
+        try:
+            self._windows.remove(weakref.ref(window))
+        except ValueError:
+            pass
+
+    def last(self) -> 'mainwindow.MainWindow':
+        """Return the last window from the access history.
+
+        Prune any None values encountered and return the last still
+        existing window from the list.
+        """
+        win = self._windows[-1]()
+        while win is None:
+            self._windows.pop()
+            win = self._windows[-1]()
+        return win
```

Comment:
WDYT about avoiding the duplication here, with something like (untested):

```suggestion
        while True:
            win = self._windows[-1]()
            if win is not None:
                return win
             self._windows.pop()
```

Also perhaps this should `return None` in case there are no windows found at all, instead of raising `IndexError`? No hard feelings either-way though.

Reply:
I agree. The only reason I left it as an IndexError was because the code already used exception handling to handle it. But returning None is cleaner. I can reduce a tiny bit more by pulling the check for `tabbed_browser.is_shutting_down` into the `WindowAccessHistory` as well.